xarray/tests/test_backends.py

```diff
@@ -3343,6 +3344,61 @@ def test_cache_members(self) -> None:
             observed_keys_2 = sorted(zstore_mut.array_keys())
             assert observed_keys_2 == sorted(array_keys + [new_key])
 
+    @requires_dask
+    @pytest.mark.parametrize("dtype", [int, float])
+    def test_zarr_fill_value_setting(self, dtype):
+        # When zarr_format=2, _FillValue sets fill_value
+        # When zarr_format=3, fill_value is set independently
+        # We test this by writing a dask array with compute=False,
+        # on read we should receive chunks filled with `fill_value`
+        fv = -1
+        ds = xr.Dataset(
+            {"foo": ("x", dask.array.from_array(np.array([0, 0, 0], dtype=dtype)))}
+        )
+        expected = xr.Dataset({"foo": ("x", [fv] * 3)})
+
+        zarr_format_2 = (
+            has_zarr_v3 and zarr.config.get("default_zarr_format") == 2
+        ) or not has_zarr_v3
+        if zarr_format_2:
+            attr = "_FillValue"
+            expected.foo.attrs[attr] = fv
+        else:
+            attr = "fill_value"
+            if dtype is float:
+                # for floats, Xarray inserts a default `np.nan`
+                expected.foo.attrs["_FillValue"] = np.nan
```

Comment:
does this assume that `np.nan` is serializable to the `attributes` field using the same JSON encoding that we use for the `fill_value` field? Because the spec doesn't cover how scalars should be encoded in `attributes`.

Reply:
Yes. I think so: https://github.com/pydata/xarray/blob/cdebbf568574a3c204ed6981c9b2660341403da9/xarray/backends/zarr.py#L1150-L1155

Reply:
since the zarr spec only defines a special JSON encoding for nan in the specific context of the `fill_value` field, I would recommend handling the JSON serialization explicitly here. 

Also, In https://github.com/zarr-developers/zarr-python/pull/2874 I remove the behavior where all numpy scalars get special serialization to JSON (instead, the dtype object is responsible for that). But it seems like we will break xarray if we move forward with that change.

Reply:
I checked what zarr v3 does today -- although we apply special JSON encoding to nan, +infinity, -inifinity when writing JSON (e.g., np.nan -> "NaN"), we only apply special decoding when reading the fill value. Is xarray basically handling that special decoding here? If so, we don't have anything to worry about w.r.t. the new dtype stuff.

Reply:
It would be great if we could take that custom attribute encoding decoding stuff out of Xarray and replace it with @d-v-b's new ZarrDType machinery.

Reply:
I have stand-alone functions that handle this, but they are currently not public API. Would be happy to make them public though!